interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,1001 @@
+		rdb.removeRuleFromTree(rule)
+		// If error occurs, rule was still fully removed from tree, and no other
+		// rule was affected. We want the rule fully removed, so this is fine.
+		rdb.notifyRule(user, rule.ID, data)
+	}
+	return nil
+}
+
+// RemoveRulesForSnapInterface removes all rules pertaining to the given snap
+// and interface for the user with the given user ID.
+func (rdb *RuleDB) RemoveRulesForSnapInterface(user uint32, snap string, iface string) ([]*Rule, error) {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	ruleFilter := func(rule *Rule) bool {
+		return rule.User == user && rule.Snap == snap && rule.Interface == iface
+	}
+	rules := rdb.rulesInternal(ruleFilter)
+	if err := rdb.removeRulesInternal(user, rules); err != nil {
+		return nil, err
+	}
+	return rules, nil
+}
+
+// PatchRule modifies the rule with the given ID by updating the rule's fields
+// corresponding to any of the given parameters which are set/non-empty.
+//
+// Any of the parameters which are equal to the default/unset value for their
+// types are left unchanged from the existing rule. Even if the given new rule
+// contents exactly match the existing rule contents, the timestamp of the rule
+// is updated to the current time. If there is any error while modifying the
+// rule, the rule is rolled back to its previous unmodified state, leaving the
+// database unchanged. If the database is changed, it is saved to disk.
+func (rdb *RuleDB) PatchRule(user uint32, id prompting.IDType, constraints *prompting.Constraints, outcome prompting.OutcomeType, lifespan prompting.LifespanType, duration string) (r *Rule, err error) {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+
+	origRule, err := rdb.ruleWithIDForUser(user, id)
+	if err != nil {
+		return nil, err
+	}
+	if constraints == nil {
+		constraints = origRule.Constraints
+	}
+	if outcome == prompting.OutcomeUnset {
+		outcome = origRule.Outcome
+	}
+	if lifespan == prompting.LifespanUnset {
+		lifespan = origRule.Lifespan
+	}
+
+	newRule, err := rdb.populateNewRule(user, origRule.Snap, origRule.Interface, constraints, outcome, lifespan, duration)
+	if err != nil {
+		return nil, err
+	}
+	newRule.ID = origRule.ID
+
+	// Remove the existing rule from the tree. An error should not occur, but
+	// if it does, carry on, hopefully the new rule won't have a conflict
+	rdb.removeRuleFromTree(origRule)
+
+	if addErr := rdb.addRuleToTree(newRule); addErr != nil {
+		err := fmt.Errorf("cannot patch rule: %w", addErr)
+		// Try to re-add original rule so all is unchanged.
+		if origErr := rdb.addRuleToTree(origRule); origErr != nil {
+			// Error should not occur, but if it does, wrap it in the other error
+			err = errorsJoin(err, fmt.Errorf("cannot re-add original rule: %w", origErr))
+			// XXX: should origRule be removed from the rules list if it's no
+			// longer in the tree?
+		}
+		return nil, err
+	}
+
+	rdb.removeRuleWithID(origRule.ID) // no error should occur, we just confirmed the rule exists
```

Comment:
it's a bit uncler given also the XXX above why we don't do this directly together with the removeRuleFromTree above?

in general I think the code should try to avoid splitting operating on the tree vs by id part as much as possible

Reply:
agree, it's pretty annoying, we're basically dealing with 2 DBs instead of just one